.specify/memory/constitution.md

```diff
@@ -0,0 +1,361 @@
+
+All TypeScript code in the dashboard repository MUST follow Backstage's architectural patterns and community best practices. Components MUST be developed in isolation using Storybook with stories demonstrating all interactive states. React components MUST follow functional component patterns with hooks. Type safety MUST be enforced through TypeScript strict mode. API clients MUST use typed interfaces generated from OpenAPI specifications. State management MUST follow Backstage plugin conventions using context and hooks appropriately.
+
+**Rationale**: The dashboard provides the primary UI for Radius users. Consistent TypeScript and React practices ensure the UI remains maintainable, accessible, and performant as the feature set grows.
+
+### XI. Frontend Testing Discipline (Dashboard)
+
+All UI components MUST include unit tests using Jest and React Testing Library, testing component behavior rather than implementation details. Interactive components MUST have Storybook stories demonstrating all states (loading, error, success, empty). Critical user workflows MUST have Playwright E2E tests covering realistic scenarios. Visual regression MUST be monitored through Storybook's visual testing tools. Accessibility MUST be validated using automated tools (axe-core) and manual keyboard navigation testing.
+
+**Rationale**: Frontend bugs directly impact user experience and are often harder to detect through manual testing. Comprehensive frontend testing ensures the dashboard remains reliable and accessible across browsers and user contexts.
+
+### XII. Resource Type Schema Quality (Contrib)
+
+All resource type schemas MUST be valid YAML files with complete property definitions, descriptions, and examples. Schemas MUST include comprehensive inline documentation explaining each field's purpose and valid values. Required vs. optional fields MUST be clearly specified. Schemas MUST follow consistent naming conventions (camelCase for properties, kebab-case for resource names). Examples MUST be runnable and demonstrate realistic usage patterns. Maturity level (Alpha, Beta, Stable) MUST be clearly documented with stability guarantees.
+
+**Rationale**: Resource type schemas define the contract for community-contributed resources. High-quality schemas ensure developers can author Radius applications confidently without ambiguity or trial-and-error.
+
+### XIII. Recipe Development Standards (Contrib)
+
+All Recipes MUST be implemented in either Bicep or Terraform with clear module structure. Recipes MUST include comprehensive README documentation explaining purpose, prerequisites, parameters, and outputs. Parameters MUST have descriptions and sensible defaults where applicable. Recipes MUST follow secure-by-default principles (e.g., disable public access, enable encryption). Recipes MUST be tested in representative environments before contribution. Recipes SHOULD be cloud-agnostic where possible, with cloud-specific variants clearly documented.
```

Comment:
what does "cloud-agnostic" mean in this context? recipes themselves by nature are platform-specific I think. a set of recipes with the same type could be cloud-agnostic though